File: backend-verification-service/src/services/ipfs.service.ts
import { Logger } from 'winston';
import { create } from 'ipfs-http-client';

interface IPFSConfig {
  apiUrl: string;
  gatewayUrl: string;
  timeout: number;
  retries: number;
  headers?: Record<string, string>;
}

interface IPFSGateway {
  url: string;
  timeout: number;
  priority: number; // Lower number = higher priority
}

export class IPFSService {
  private ipfs: any;
  private gateways: IPFSGateway[];
  private currentGatewayIndex: number = 0;
  private maxRetries: number = 3;
  private retryDelay: number = 1000; // 1 second
  private config: IPFSConfig;

  constructor(private logger: Logger) {
    // Initialize configuration
    this.config = {
      apiUrl: process.env.IPFS_API_URL || 'https://ipfs.infura.io:5001/api/v0',
      gatewayUrl: process.env.IPFS_GATEWAY_URL || 'https://ipfs.io',
      timeout: parseInt(process.env.IPFS_TIMEOUT || '30000'), // 30 seconds
      retries: parseInt(process.env.IPFS_RETRIES || '3')
    };

    // Multiple IPFS gateways for fallback (ordered by priority)
    this.gateways = [
      { url: this.config.gatewayUrl, timeout: 5000, priority: 1 },
      { url: 'https://gateway.pinata.cloud', timeout: 8000, priority: 2 },
      { url: 'https://cloudflare-ipfs.com', timeout: 10000, priority: 3 },
      { url: 'https://dweb.link', timeout: 10000, priority: 4 },
      { url: 'https://ipfs.fleek.co', timeout: 12000, priority: 5 },
      { url: 'https://gateway.ipfs.io', timeout: 15000, priority: 6 }
    ];

    // Sort gateways by priority
    this.gateways.sort((a, b) => a.priority - b.priority);
    
    // Initialize IPFS client with authentication if available
    const clientOptions: any = { 
      url: this.config.apiUrl,
      timeout: this.config.timeout
    };

    // Add authentication headers if configured
    if (process.env.IPFS_PROJECT_ID && process.env.IPFS_API_SECRET) {
      const auth = 'Basic ' + Buffer.from(
        process.env.IPFS_PROJECT_ID + ':' + process.env.IPFS_API_SECRET
      ).toString('base64');
      
      clientOptions.headers = {
        authorization: auth
      };
    }

    this.ipfs = create(clientOptions);
  }

  async initialize(): Promise<void> {
    try {
      // Test IPFS connection
      await this.testConnection();
      
      // Test gateway connectivity
      await this.testGateways();
      
      this.logger.info('IPFS service initialized successfully', {
        apiUrl: this.config.apiUrl,
        gatewayCount: this.gateways.length,
        primaryGateway: this.gateways[0]?.url || 'unknown'
      });
    } catch (error) {
      this.logger.error('IPFS service initialization failed:', error);
      throw error;
    }
  }

  async healthCheck(): Promise<boolean> {
    try {
      // Test basic IPFS functionality
      const testCid = await this.storeDocument('health check test');
      if (!testCid) {
        return false;
      }

      // Test retrieval
      const retrieved = await this.retrieveDocument(testCid);
      return retrieved === 'health check test';
    } catch (error) {
      this.logger.error('IPFS service health check failed:', error);
      return false;
    }
  }

  private async testConnection(): Promise<void> {
    try {
      // Test IPFS API connection
      const version = await this.ipfs.version();
      this.logger.debug('IPFS API connection successful', { version: version.version });
    } catch (error) {
      throw new Error(`IPFS API connection failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  private async testGateways(): Promise<void> {
    const workingGateways: IPFSGateway[] = [];
    
    // Test each gateway with a known IPFS hash
    const testCid = 'QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG'; // "hello world" on IPFS
    
    for (const gateway of this.gateways) {
      try {
        const url = `${gateway.url}/ipfs/${testCid}`;
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), gateway.timeout);
        
        const response = await fetch(url, { 
          method: 'HEAD',
          signal: controller.signal
        });
        
        clearTimeout(timeoutId);
        
        if (response.ok) {
          workingGateways.push(gateway);
          this.logger.debug('Gateway test successful', { gateway: gateway.url });
        } else {
          this.logger.warn('Gateway test failed', { 
            gateway: gateway.url, 
            status: response.status 
          });
        }
      } catch (error) {
        this.logger.warn('Gateway test error', { 
          gateway: gateway.url, 
          error: error instanceof Error ? error.message : 'Unknown error'
        });
      }
    }
    
    if (workingGateways.length === 0) {
      this.logger.warn('No working IPFS gateways found, using original list');
    } else {
      // Update gateway list to prioritize working ones
      this.gateways = [...workingGateways, ...this.gateways.filter(g => !workingGateways.includes(g))];
      this.logger.info('Gateway connectivity test completed', { 
        workingGateways: workingGateways.length,
        totalGateways: this.gateways.length 
      });
    }
  }

  async storeDocument(content: string | Buffer): Promise<string> {
    let lastError: Error | null = null;
    
    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        this.logger.debug('Storing document on IPFS', { 
          attempt: attempt + 1, 
          contentSize: content.length 
        });

        const result = await this.ipfs.add(content, {
          pin: true, // Pin the content to ensure it stays available
          cidVersion: 1, // Use CIDv1 for better compatibility
          hashAlg: 'sha2-256'
        });
        
        const cid = result.cid.toString();
        
        this.logger.info('Document stored successfully on IPFS', { 
          cid,
          size: result.size,
          attempt: attempt + 1
        });

        // Verify the upload by attempting to retrieve metadata
        await this.validateStoredDocument(cid);
        
        return cid;
      } catch (error) {
        lastError = error as Error;
        this.logger.warn(`IPFS store attempt ${attempt + 1} failed:`, {
          error: lastError.message,
          attempt: attempt + 1,
          maxRetries: this.maxRetries
        });
        
        if (attempt < this.maxRetries - 1) {
          const delay = this.retryDelay * Math.pow(2, attempt); // Exponential backoff
          await this.delay(delay);
        }
      }
    }
    
    this.logger.error('Failed to store document in IPFS after all retries:', lastError);
    throw lastError || new Error('IPFS store operation failed');
  }

  async retrieveDocument(cid: string): Promise<string> {
    let lastError: Error | null = null;
    
    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        this.logger.debug('Retrieving document from IPFS', { cid, attempt: attempt + 1 });

        const chunks = [];
        for await (const chunk of this.ipfs.cat(cid, { timeout: this.config.timeout })) {
          chunks.push(chunk);
        }
        
        const content = Buffer.concat(chunks).toString();
        
        this.logger.debug('Document retrieved successfully from IPFS', { 
          cid, 
          size: content.length,
          attempt: attempt + 1
        });
        
        return content;
      } catch (error) {
        lastError = error as Error;
        this.logger.warn(`IPFS retrieve attempt ${attempt + 1} failed for CID ${cid}:`, {
          error: lastError.message,
          attempt: attempt + 1
        });
        
        if (attempt < this.maxRetries - 1) {
          const delay = this.retryDelay * Math.pow(2, attempt);
          await this.delay(delay);
        }
      }
    }
    
    this.logger.error(`Failed to retrieve document ${cid} from IPFS after all retries:`, lastError);
    throw lastError || new Error(`IPFS retrieve operation failed for CID: ${cid}`);
  }

  private async validateStoredDocument(cid: string): Promise<boolean> {
    try {
      // Try to get the object stats to verify it exists
      const stats = await this.ipfs.object.stat(cid, { timeout: 5000 });
      return stats && stats.Hash === cid;
    } catch (error) {
      this.logger.warn('Document validation failed', { cid, error: error instanceof Error ? error.message : 'Unknown error' });
      return false;
    }
  }

  async getGatewayUrl(cid: string): Promise<string> {
    return `${this.gateways[this.currentGatewayIndex]?.url || 'https://ipfs.io'}/ipfs/${cid}`;
  }

  async getGatewayUrlWithFallback(cid: string): Promise<string> {
    // Try to find a working gateway for this specific CID
    for (let i = 0; i < this.gateways.length; i++) {
      const gateway = this.gateways[i];
      if (!gateway) continue;
      
      const gatewayUrl = `${gateway.url}/ipfs/${cid}`;
      
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), gateway.timeout);
        
        const response = await fetch(gatewayUrl, { 
          method: 'HEAD',
          signal: controller.signal
        });
        
        clearTimeout(timeoutId);
        
        if (response.ok) {
          this.currentGatewayIndex = i; // Update to working gateway
          this.logger.debug('Gateway availability verified', { gateway: gateway.url, cid });
          return gatewayUrl;
        }
      } catch (error) {
        this.logger.debug('Gateway check failed', { 
          gateway: gateway.url, 
          cid,
          error: error instanceof Error ? error.message : 'Unknown error'
        });
        continue;
      }
    }
    
    // If all gateways fail, return the primary gateway URL
    this.logger.warn(`All IPFS gateways failed for CID ${cid}, using primary gateway`);
    return `${this.gateways[0]?.url || 'https://ipfs.io'}/ipfs/${cid}`;
  }

  async validateCID(cid: string): Promise<boolean> {
    try {
      // Basic CID format validation
      if (!cid || typeof cid !== 'string') {
        return false;
      }

      // Try to retrieve object stats to validate CID exists
      const stats = await this.ipfs.object.stat(cid, { timeout: 10000 });
      return !!stats;
    } catch (error) {
      this.logger.debug(`CID validation failed for ${cid}:`, error instanceof Error ? error.message : 'Unknown error');
      return false;
    }
  }

  async pinDocument(cid: string): Promise<boolean> {
    try {
      await this.ipfs.pin.add(cid);
      this.logger.info('Document pinned successfully', { cid });
      return true;
    } catch (error) {
      this.logger.error('Failed to pin document:', { cid, error: error instanceof Error ? error.message : 'Unknown error' });
      return false;
    }
  }

  async unpinDocument(cid: string): Promise<boolean> {
    try {
      await this.ipfs.pin.rm(cid);
      this.logger.info('Document unpinned successfully', { cid });
      return true;
    } catch (error) {
      this.logger.error('Failed to unpin document:', { cid, error: error instanceof Error ? error.message : 'Unknown error' });
      return false;
    }
  }

  async getDocumentStats(cid: string): Promise<{
    size: number;
    cumulativeSize: number;
    blocks: number;
    type: string;
  } | null> {
    try {
      const stats = await this.ipfs.object.stat(cid);
      return {
        size: stats.DataSize || 0,
        cumulativeSize: stats.CumulativeSize || 0,
        blocks: stats.NumLinks || 0,
        type: 'object'
      };
    } catch (error) {
      this.logger.error('Failed to get document stats:', { cid, error: error instanceof Error ? error.message : 'Unknown error' });
      return null;
    }
  }

  getGatewayStatus(): { 
    currentGateway: string; 
    availableGateways: IPFSGateway[];
    config: IPFSConfig;
  } {
    return {
      currentGateway: this.gateways[this.currentGatewayIndex]?.url || 'none',
      availableGateways: this.gateways,
      config: this.config
    };
  }

  async getServiceStats(): Promise<{
    version?: string;
    peerId?: string;
    repo?: any;
    isOnline: boolean;
    gatewayCount: number;
    currentGateway: string;
  }> {
    try {
      const [version, id] = await Promise.allSettled([
        this.ipfs.version(),
        this.ipfs.id()
      ]);

      return {
        version: version.status === 'fulfilled' ? version.value.version : undefined,
        peerId: id.status === 'fulfilled' ? id.value.id : undefined,
        isOnline: true,
        gatewayCount: this.gateways.length,
        currentGateway: this.gateways[this.currentGatewayIndex]?.url || 'unknown'
      };
    } catch (error) {
      this.logger.error('Failed to get IPFS service stats:', error);
      return {
        isOnline: false,
        gatewayCount: this.gateways.length,
        currentGateway: 'unknown'
      };
    }
  }

  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
} 

File: data-pipeline/src/veritas_data_pipeline/services/ipfs_service.py
"""
IPFS Service for Veritas Data Pipeline
Handles document storage and retrieval with multi-gateway fallback
"""
import os
import asyncio
import aiohttp
import structlog
from typing import Optional, Dict, Any
from ipfshttpclient import connect, Client

logger = structlog.get_logger()

class IPFSService:
    """IPFS service with multi-gateway fallback support"""
    
    def __init__(self):
        # Multiple IPFS gateways for fallback
        self.gateway_urls = [
            os.getenv('IPFS_GATEWAY_URL', 'https://ipfs.io'),
            'https://gateway.pinata.cloud',
            'https://cloudflare-ipfs.com',
            'https://dweb.link',
            'https://ipfs.fleek.co'
        ]
        
        # IPFS API endpoints
        self.api_urls = [
            os.getenv('IPFS_API_URL', 'https://ipfs.infura.io:5001/api/v0'),
            'https://ipfs.io:5001/api/v0',
            'https://gateway.pinata.cloud:5001/api/v0'
        ]
        
        self.current_gateway_index = 0
        self.current_api_index = 0
        self.max_retries = 3
        self.retry_delay = 1.0  # 1 second
        self.client: Optional[Client] = None
        
    async def initialize(self) -> None:
        """Initialize IPFS client with fallback"""
        try:
            # Try to connect to IPFS API endpoints
            for i, api_url in enumerate(self.api_urls):
                try:
                    self.client = connect(api_url, session=True)
                    self.current_api_index = i
                    logger.info("IPFS client initialized", api_url=api_url)
                    break
                except Exception as e:
                    logger.warning("Failed to connect to IPFS API", api_url=api_url, error=str(e))
                    continue
            
            if not self.client:
                raise Exception("Failed to connect to any IPFS API endpoint")
            
            # Test connection
            test_cid = await self.store_document("test")
            if not test_cid:
                raise Exception("Failed to store test document")
                
            logger.info("IPFS service initialized successfully with multi-gateway support")
            
        except Exception as e:
            logger.error("IPFS service initialization failed", error=str(e))
            raise
    
    async def health_check(self) -> bool:
        """Check IPFS service health"""
        try:
            test_cid = await self.store_document("health check")
            return bool(test_cid)
        except Exception as e:
            logger.error("IPFS health check failed", error=str(e))
            return False
    
    async def store_document(self, content: str) -> str:
        """Store document in IPFS with retry logic"""
        if not self.client:
            raise Exception("IPFS client not initialized")
        
        last_error = None
        
        for attempt in range(self.max_retries):
            try:
                # Try current API endpoint
                result = self.client.add_str(content)
                return result
            except Exception as e:
                last_error = e
                logger.warning("IPFS store attempt failed", 
                             attempt=attempt + 1, 
                             api_url=self.api_urls[self.current_api_index],
                             error=str(e))
                
                if attempt < self.max_retries - 1:
                    # Try next API endpoint
                    self.current_api_index = (self.current_api_index + 1) % len(self.api_urls)
                    try:
                        self.client = connect(self.api_urls[self.current_api_index], session=True)
                        await asyncio.sleep(self.retry_delay * (2 ** attempt))  # Exponential backoff
                    except Exception as reconnect_error:
                        logger.warning("Failed to reconnect to IPFS API", 
                                     api_url=self.api_urls[self.current_api_index],
                                     error=str(reconnect_error))
                        continue
        
        logger.error("Failed to store document in IPFS after all retries", error=str(last_error))
        raise last_error or Exception("IPFS store operation failed")
    
    async def retrieve_document(self, cid: str) -> str:
        """Retrieve document from IPFS with retry logic"""
        if not self.client:
            raise Exception("IPFS client not initialized")
        
        last_error = None
        
        for attempt in range(self.max_retries):
            try:
                # Try current API endpoint
                result = self.client.cat(cid)
                return result.decode('utf-8')
            except Exception as e:
                last_error = e
                logger.warning("IPFS retrieve attempt failed", 
                             attempt=attempt + 1, 
                             cid=cid,
                             api_url=self.api_urls[self.current_api_index],
                             error=str(e))
                
                if attempt < self.max_retries - 1:
                    # Try next API endpoint
                    self.current_api_index = (self.current_api_index + 1) % len(self.api_urls)
                    try:
                        self.client = connect(self.api_urls[self.current_api_index], session=True)
                        await asyncio.sleep(self.retry_delay * (2 ** attempt))
                    except Exception as reconnect_error:
                        logger.warning("Failed to reconnect to IPFS API", 
                                     api_url=self.api_urls[self.current_api_index],
                                     error=str(reconnect_error))
                        continue
        
        logger.error("Failed to retrieve document from IPFS after all retries", 
                    cid=cid, error=str(last_error))
        raise last_error or Exception(f"IPFS retrieve operation failed for CID: {cid}")
    
    def get_gateway_url(self, cid: str) -> str:
        """Get gateway URL for CID"""
        return f"{self.gateway_urls[self.current_gateway_index]}/ipfs/{cid}"
    
    async def get_gateway_url_with_fallback(self, cid: str) -> str:
        """Get working gateway URL with fallback"""
        async with aiohttp.ClientSession() as session:
            for i, gateway_url in enumerate(self.gateway_urls):
                full_url = f"{gateway_url}/ipfs/{cid}"
                
                try:
                    async with session.head(full_url, timeout=aiohttp.ClientTimeout(total=5)) as response:
                        if response.status == 200:
                            self.current_gateway_index = i
                            return full_url
                except Exception as e:
                    logger.debug("Gateway failed", gateway=gateway_url, cid=cid, error=str(e))
                    continue
            
            # If all gateways fail, return primary gateway URL
            logger.warning("All IPFS gateways failed", cid=cid)
            return f"{self.gateway_urls[0]}/ipfs/{cid}"
    
    async def validate_cid(self, cid: str) -> bool:
        """Validate CID by attempting to retrieve a small portion"""
        try:
            if not self.client:
                return False
                
            # Try to retrieve first chunk
            result = self.client.cat(cid, length=100)  # Get first 100 bytes
            return len(result) > 0
        except Exception as e:
            logger.error("CID validation failed", cid=cid, error=str(e))
            return False
    
    def get_gateway_status(self) -> Dict[str, Any]:
        """Get current gateway status"""
        return {
            "current_gateway": self.gateway_urls[self.current_gateway_index],
            "current_api": self.api_urls[self.current_api_index],
            "available_gateways": self.gateway_urls,
            "available_apis": self.api_urls
        }
    
    async def close(self) -> None:
        """Close IPFS client connection"""
        if self.client:
            try:
                self.client.close()
                logger.info("IPFS client connection closed")
            except Exception as e:
                logger.warning("Error closing IPFS client", error=str(e)) 